### PR TITLE
Binary (0-1) Bernoulli-Gaussian RBM

### PR DESCRIPTION
**Pull request by Francill99**

### CODE REVIEW

File: rbms/bernoulli_gaussian/classes.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+        if device is None: device = weight_matrix.device
+        if dtype is None: dtype = weight_matrix.dtype
+        self.device, self.dtype = device, dtype
+
+        self.weight_matrix = weight_matrix.to(device=self.device, dtype=self.dtype)
+        self.vbias = vbias.to(device=self.device, dtype=self.dtype)
+        self.hbias = hbias.to(device=self.device, dtype=self.dtype)
+
+        Nv = int(self.vbias.numel())
+        Nh = int(self.hbias.numel())
```
Comment (nbereux) [inline rbms/bernoulli_gaussian/classes.py]: These should be removed I think, they are not useful and redundant with `self.num_visibles()` and `self.num_hiddens()`

File: rbms/bernoulli_gaussian/classes.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+            v=v, vbias=self.vbias, hbias=self.hbias, weight_matrix=self.weight_matrix
+        )
+
+    def compute_gradient(
+        self, data, chains, centered=True, lambda_l1=0.0, lambda_l2=0.0
+    ):
+        # backend should ignore grads on eta or treat it as const; we pass it for conditionals
+        _compute_gradient(
+            v_data=data["visible"],
+            h_data=data["hidden"],
```
Comment (nbereux) [inline rbms/bernoulli_gaussian/classes.py]: should be `data["hidden_mag"]`

File: rbms/bernoulli_gaussian/classes.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+            w_data=data["weights"],
+            v_chain=chains["visible"],
+            h_chain=chains["hidden"],
```
Comment (nbereux) [inline rbms/bernoulli_gaussian/classes.py]: should be `chains["hidden_mag"]`

File: rbms/bernoulli_gaussian/functional.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    centered: bool,
+    lambda_l1: float = 0.0,
+    lambda_l2: float = 0.0,
+) -> None:
+    _compute_gradient(
+        v_data=data["visible"],
+        mh_data=data["hidden_mag"],   # use conditional mean for positive phase
+        w_data=data["weights"],
+        v_chain=chains["visible"],
+        h_chain=chains["hidden"],     # negative phase from chain samples
```
Comment (nbereux) [inline rbms/bernoulli_gaussian/functional.py]: should be `chains["hidden_mag"]`

File: rbms/bernoulli_gaussian/implement.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+from torch import Tensor
+from torch.nn.functional import softmax
+
+
+@torch.jit.script
+def _sample_hiddens(
+    v: Tensor, weight_matrix: Tensor, hbias: Tensor, beta: float = 1.0
+) -> Tuple[Tensor, Tensor]:
+    mh = (hbias + (v @ weight_matrix))
+    h = torch.normal(mean=mh, std=torch.tensor(1.0).to(weight_matrix.device))
```
Comment (nbereux) [inline rbms/bernoulli_gaussian/implement.py]: I would recommend to do `h = torch.randn_like(mh) + mh`. This avoids synchronization between CPU/GPU at every sampling steps and the creation + cast of the `std` tensor at every step.

File: rbms/bernoulli_gaussian/implement.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    weight_matrix: Tensor,
+) -> Tensor:
+    fields = torch.tensordot(vbias, v, dims=[[0], [1]]) + torch.tensordot(
+        hbias, h, dims=[[0], [1]]
+    )
+    interaction = torch.multiply(
+        v, torch.tensordot(h, weight_matrix, dims=[[1], [1]])
+    ).sum(1)
+    Nv = weight_matrix.shape[0]
+    gamma = float(Nv)
```
Comment (nbereux) [inline rbms/bernoulli_gaussian/implement.py]: Why do you need this cast ?

File: rbms/bernoulli_gaussian/implement.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+    field = v @ vbias  # (B,)
+    t = hbias + (v @ weight_matrix)  # (B,K)
+    Nv = weight_matrix.shape[0]
+    K = weight_matrix.shape[1]
+    inv_gamma = 1.0 / float(Nv)
+    quad_term = 0.5 * inv_gamma * (t * t).sum(1)  # (B,)
+    dtype = v.dtype
+    device = v.device
+    log_two_pi = torch.log(torch.tensor(2.0 * torch.pi, dtype=dtype, device=device))
+    const = 0.5 * float(K) * (torch.log(torch.tensor(float(Nv), dtype=dtype, device=device)) - log_two_pi)
```
Comment (nbereux) [inline rbms/bernoulli_gaussian/implement.py]: I think we need to benchmark this to see if there is a more efficient way to do it. because the construction of tensor might be expensive.

**Review by nbereux — CHANGES_REQUESTED** [review]
I put some comments but overall it seems very nice ! 
I would also request that you format the code with [ruff](https://docs.astral.sh/ruff/) afterwards. The formatting rules are set in the `pyproject.toml`

Also you can modify it to add yourself as contributor.

File: rbms/bernoulli_gaussian/implement.py (hunk 6 above)
Comment (Francill99) [reply rbms/bernoulli_gaussian/implement.py]: At first I left open the possibility to change gamma, that means that you could select the variance of the prior over Gaussian hiddens. Then for simplicity I passed to a fixed variance equal to 1/N_v as the common choice for statphys computations and it remained. I told you instead of eliminating it directly because I still have the question: do we leave the possibility to select the variance or do we fix it once for all?

File: rbms/bernoulli_gaussian/functional.py (hunk 4 above)
Comment (Francill99) [reply rbms/bernoulli_gaussian/functional.py]: I will try if, with this change, the training is stable for a higher learning rate (maybe this was the problem)

File: rbms/bernoulli_gaussian/implement.py (hunk 7 above)
Comment (Francill99) [reply rbms/bernoulli_gaussian/implement.py]: Now I compute the constant once for all as self.const in classes and I pass it as an input to compute_energy method

File: rbms/bernoulli_gaussian/implement.py (hunk 6 above)
Comment (nbereux) [reply rbms/bernoulli_gaussian/implement.py]: Your call there, have you seen anyone using a different variance ? I don't think it would change anything... or maybe if you want non uniform variance across the hidden nodes for obscure reasons...

Comment (Francill99) [reply rbms/bernoulli_gaussian/implement.py]: Okay now I am convinced: if the variance is set equal for all hidden variables then it simply rescales model effective temperature. So the only case where it is useful is if hiddens have different variance, but I don't imagine a case where it is of interest. Then I simply removed gamma definition

**Comment by Francill99** [comment]
I solved all requests and formatted the code using rust. Now, how to send again the code modified? Do I make a new merge, do I simply commit on my forked main, or is there another way of updating correctly?

**Comment by nbereux** [comment]
Yes you should commit to the branch you were working on before and push, the changes will appear here !

**Comment by Francill99** [comment]
I committed the new version with all changes requested. For any other issue don't esitate to contact me. I tried some trainings with higher values of learning rate with new version but it seems to need still lower ones, as well as long mcmc for sampling after training

File: rbms/bernoulli_gaussian/classes.py [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+    def compute_energy_visibles(self, v: Tensor) -> Tensor:
+        return _compute_energy_visibles(
+            v=v, vbias=self.vbias, hbias=self.hbias, weight_matrix=self.weight_matrix
+        )
+
+    def compute_gradient(self, data, chains, centered=True, lambda_l1=0.0, lambda_l2=0.0):
+        # backend should ignore grads on eta or treat it as const; we pass it for conditionals
+        _compute_gradient(
+            v_data=data["visible"],
+            h_data=data["hidden_magn"],
```
Comment (nbereux) [inline rbms/bernoulli_gaussian/classes.py]: Typo: `"hidden_magn"` -> `"hidden_mag"`

File: rbms/bernoulli_gaussian/classes.py [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+        )
+
+    def compute_energy_hiddens(self, h: Tensor) -> Tensor:
+        return _compute_energy_hiddens(
+            h=h, vbias=self.vbias, hbias=self.hbias, weight_matrix=self.weight_matrix
+        )
+
+    def compute_energy_visibles(self, v: Tensor) -> Tensor:
+        return _compute_energy_visibles(
+            v=v, vbias=self.vbias, hbias=self.hbias, weight_matrix=self.weight_matrix
```
Comment (nbereux) [inline rbms/bernoulli_gaussian/classes.py]: missing: `const=self.const`

File: rbms/bernoulli_gaussian/implement.py [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+    weight_matrix: Tensor,
+) -> Tensor:
+    fields = torch.tensordot(vbias, v, dims=[[0], [1]]) + torch.tensordot(
+        hbias, h, dims=[[0], [1]]
+    )
+    interaction = torch.multiply(
+        v, torch.tensordot(h, weight_matrix, dims=[[1], [1]])
+    ).sum(1)
+    Nv = weight_matrix.shape[0]
+    quad = 0.5 * Nv * (h * h).sum(1)
```
Comment (nbereux) [inline rbms/bernoulli_gaussian/implement.py]: `Nv` -> `num_visibles` to be consistent in the names of variables of the repo

File: rbms/bernoulli_gaussian/implement.py [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+
+@torch.jit.script
+def _compute_energy_visibles(
+    v: Tensor, vbias: Tensor, hbias: Tensor, weight_matrix: Tensor, const: Tensor,
+) -> Tensor:
+    field = v @ vbias  
+    t = hbias + (v @ weight_matrix) 
+    Nv = weight_matrix.shape[0]
+    inv_gamma = 1.0 / Nv
+    quad_term = 0.5 * inv_gamma * (t * t).sum(1)
```
Comment (nbereux) [inline rbms/bernoulli_gaussian/implement.py]: `Nv` -> `num_visibles` to be consistent in the names of variables of the repo.
Also remove `inv_gamma = 1.0 / Nv` and do `quad_term = 0.5 *  (t * t).sum(1) / num_visibles` to remove the `gamma` notation

File: rbms/bernoulli_gaussian/implement.py [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+@torch.jit.script
+def _compute_energy_hiddens(
+    h: Tensor, vbias: Tensor, hbias: Tensor, weight_matrix: Tensor
+) -> Tensor:
+    field = h @ hbias 
+    exponent = vbias + (h @ weight_matrix.T) 
+    log_term = torch.where(exponent < 10, torch.log1p(torch.exp(exponent)), exponent)
+    Nv = weight_matrix.shape[0]
+    gamma = float(Nv)
+    quad = 0.5 * gamma * (h * h).sum(1)
```
Comment (nbereux) [inline rbms/bernoulli_gaussian/implement.py]: `Nv` -> `num_visibles` to be consistent in the names of variables of the repo.
Also remove `gamma = Nv` and do `quad_term = 0.5 * num_visibles * (t * t).sum(1)` to remove the `gamma` notation

**Review by nbereux — CHANGES_REQUESTED** [review]
There are some typos that need to be fixed and then all good for merging !

**Comment by nbereux** [comment]
Also put your name as contributor in the `pyproject.toml` !!

**Review by nbereux — APPROVED** [review]
